sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala

```diff
@@ -1265,4 +1265,168 @@ class DataFrameWindowFunctionsSuite extends QueryTest
+            )
+          )
+
+          checkAnswer(df.withColumn("rn", rank().over(window)).where(condition),
+            Seq(
+              Row("a", 4, "", 2.0, 1),
+              Row("a", 4, "", 2.0, 1),
+              Row("b", 1, "h", Double.NaN, 1),
+              Row("b", 1, "n", Double.PositiveInfinity, 2),
+              Row("c", 1, "a", -4.0, 2),
+              Row("c", 2, null, 5.0, 1)
+            )
+          )
+
+          checkAnswer(df.withColumn("rn", dense_rank().over(window)).where(condition),
+            Seq(
+              Row("a", 0, "c", 1.0, 2),
+              Row("a", 4, "", 2.0, 1),
+              Row("a", 4, "", 2.0, 1),
+              Row("b", 1, "h", Double.NaN, 1),
+              Row("b", 1, "n", Double.PositiveInfinity, 2),
+              Row("c", 1, "a", -4.0, 2),
+              Row("c", 2, null, 5.0, 1)
+            )
+          )
+        }
+
+        val condition = $"rn" === 2 && $"value2" > 0.5
+        checkAnswer(df.withColumn("rn", row_number().over(window)).where(condition),
+          Seq(
+            Row("a", 4, "", 2.0, 2),
+            Row("b", 1, "n", Double.PositiveInfinity, 2)
+          )
+        )
+
+        checkAnswer(df.withColumn("rn", rank().over(window)).where(condition),
+          Seq(
+            Row("b", 1, "n", Double.PositiveInfinity, 2)
+          )
+        )
+
+        checkAnswer(df.withColumn("rn", dense_rank().over(window)).where(condition),
+          Seq(
+            Row("a", 0, "c", 1.0, 2),
+            Row("b", 1, "n", Double.PositiveInfinity, 2)
+          )
+        )
+
+        val multipleRowNumbers = df
+          .withColumn("rn", row_number().over(window))
+          .withColumn("rn2", row_number().over(window))
+          .where('rn < 2 && 'rn2 < 3)
+        checkAnswer(multipleRowNumbers,
+          Seq(
+            Row("a", 4, "", 2.0, 1, 1),
+            Row("b", 1, "h", Double.NaN, 1, 1),
+            Row("c", 2, null, 5.0, 1, 1)
+          )
+        )
+
+        val multipleRanks = df
+          .withColumn("rn", rank().over(window))
+          .withColumn("rn2", rank().over(window))
+          .where('rn < 2 && 'rn2 < 3)
+        checkAnswer(multipleRanks,
+          Seq(
+            Row("a", 4, "", 2.0, 1, 1),
+            Row("a", 4, "", 2.0, 1, 1),
+            Row("b", 1, "h", Double.NaN, 1, 1),
+            Row("c", 2, null, 5.0, 1, 1)
+          )
+        )
+
+        val multipleDenseRanks = df
+          .withColumn("rn", dense_rank().over(window))
+          .withColumn("rn2", dense_rank().over(window))
+          .where('rn < 2 && 'rn2 < 3)
+        checkAnswer(multipleDenseRanks,
+          Seq(
+            Row("a", 4, "", 2.0, 1, 1),
+            Row("a", 4, "", 2.0, 1, 1),
+            Row("b", 1, "h", Double.NaN, 1, 1),
+            Row("c", 2, null, 5.0, 1, 1)
+          )
+        )
+
+        val multipleWindowsOne = df
+          .withColumn("rn2", row_number().over(window2))
+          .withColumn("rn", row_number().over(window))
```

Comment:
How does this work? If there are two window specs, then the query has two Window node?

Reply:
Yes. The inner Window will not be optimized.